### PR TITLE
Support weaviate

### PR DESCRIPTION
**Pull request by xucailiang**
[Issues](https://github.com/chonkie-inc/chonkie/issues/203)

Test:
```
import os
from chonkie import OpenAIEmbeddings, RecursiveChunker, WeaviateHandshake

os.environ['OPENAI_API_KEY'] = "your api key"

embeddings = OpenAIEmbeddings()

# Initialize the chunker
chunker = RecursiveChunker()

handshake = WeaviateHandshake(embedding_model=embeddings)

# Chunk some text
chunks = chunker("Chonkie is the goodest boi! My favorite chunking hippo hehe.")

# Access chunks
for chunk in chunks:
    print(f"{chunk.__dict__}")
    print(f"Chunk: {chunk.text}")
    print(f"Tokens: {chunk.token_count}")

    embed_list = handshake.embedding_model.embed(chunk.text)
    print(embed_list)

results = handshake.write(chunks=chunks)
print(results)

handshake.close()
```

docker-compose-weaviate.yaml:
```
services:
  weaviate:
    command:
    - --host
    - 0.0.0.0
    - --port
    - '8080'
    - --scheme
    - http
    image: cr.weaviate.io/semitechnologies/weaviate:1.31.3
    ports:
    - 8080:8080
    - 50051:50051
    volumes:
    - weaviate_data:/var/lib/weaviate
    restart: on-failure:0
    environment:
      QUERY_DEFAULTS_LIMIT: 25
      AUTHENTICATION_ANONYMOUS_ACCESS_ENABLED: 'true'
      PERSISTENCE_DATA_PATH: '/var/lib/weaviate'
      ENABLE_API_BASED_MODULES: 'true'
      CLUSTER_HOSTNAME: 'node1'
volumes:
  weaviate_data:
```

### CODE REVIEW

File: src/chonkie/friends/handshakes/weaviate.py [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+
+    def _import_dependencies(self) -> None:
+        """Lazy import the dependencies."""
+        if self._is_available():
+            global weaviate
+            import weaviate
+        else:
+            raise ImportError(
+                "Weaviate client is not installed. "
+                "Please install it with `pip install chonkie[weaviate-client]`."
```
Comment (chonk-lain) [inline src/chonkie/friends/handshakes/weaviate.py]: add support for this `weaviate-client` in the pyproject.toml in the `project.optional-dependencies` field

File: src/chonkie/friends/handshakes/weaviate.py [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+            self.client = weaviate.connect_to_custom(
+                http_host=host,
+                http_port=port,
+                http_secure=http_secure,
+                grpc_host=actual_grpc_host,
+                grpc_port=grpc_port,
+                grpc_secure=grpc_secure,
+                auth_credentials=auth_credentials,
+                headers=additional_headers
+            )
```
Comment (chonk-lain) [inline src/chonkie/friends/handshakes/weaviate.py]: if i am manually passing my url and api key this section it does not work.
a minimalistic reproducer would be 

```python
from chonkie import WeaviateHandshake

handshake = WeaviateHandshake(url=weaviate_url, api_key=WEAVIATE_KEY) # I generated both of these from https://console.weaviate.cloud/ 
```
```
UnexpectedStatusCodeError                 Traceback (most recent call last)
[/tmp/ipython-input-2351092521.py](https://localhost:8080/#) in <cell line: 0>()
----> 1 handshake = WeaviateHandshake(url=weaviate_url, api_key=WEAVIATE_KEY)

14 frames
[/usr/local/lib/python3.11/dist-packages/chonkie/friends/handshakes/weaviate.py](https://localhost:8080/#) in __init__(self, client, collection_name, embedding_model, url, api_key, auth_config, batch_size, batch_dynamic, batch_timeout_retries, additional_headers, http_secure, grpc_host, grpc_port, grpc_secure)
    105             actual_grpc_host = grpc_host if grpc_host is not None else host
    106 
--> 107             self.client = weaviate.connect_to_custom(
    108                 http_host=host,
    109                 http_port=port,

[/usr/local/lib/python3.11/dist-packages/weaviate/connect/helpers.py](https://localhost:8080/#) in connect_to_custom(http_host, http_port, http_secure, grpc_host, grpc_port, grpc_secure, headers, additional_config, auth_credentials, skip_init_checks)
    345         >>> # The connection is automatically closed when the context is exited.
    346     """
--> 347     return __connect(
    348         WeaviateClient(
    349             ConnectionParams.from_params(

[/usr/local/lib/python3.11/dist-packages/weaviate/connect/helpers.py](https://localhost:8080/#) in __connect(client)
    369     except Exception as e:
    370         client.close()
--> 371         raise e
    372 
    373 

[/usr/local/lib/python3.11/dist-packages/weaviate/connect/helpers.py](https://localhost:8080/#) in __connect(client)
    365 def __connect(client: WeaviateClient) -> WeaviateClient:
    366     try:
--> 367         client.connect()
    368         return client
    369     except Exception as e:

[/usr/local/lib/python3.11/dist-packages/weaviate/client_executor.py](https://localhost:8080/#) in connect(self)
    147             weaviate.exceptions.UnexpectedStatusCodeError: If weaviate reports a none OK status.
    148         """
--> 149         return executor.execute(
    150             response_callback=lambda _: None,
    151             method=self._connection.connect,

[/usr/local/lib/python3.11/dist-packages/weaviate/connect/executor.py](https://localhost:8080/#) in execute(method, response_callback, exception_callback, *args, **kwargs)
     97         return resp_call
     98     except Exception as e:
---> 99         return cast(T, exception_callback(e))
    100 
    101 

[/usr/local/lib/python3.11/dist-packages/weaviate/connect/executor.py](https://localhost:8080/#) in raise_exception(e)
     36 
     37 def raise_exception(e: Exception) -> Any:
---> 38     raise e
     39 
     40 

[/usr/local/lib/python3.11/dist-packages/weaviate/connect/executor.py](https://localhost:8080/#) in execute(method, response_callback, exception_callback, *args, **kwargs)
     78     # wrap method call in try-except to catch exceptions for sync method
     79     try:
---> 80         call = method(*args, **kwargs)
     81         if isinstance(call, Awaitable):
     82 

[/usr/local/lib/python3.11/dist-packages/weaviate/connect/v4.py](https://localhost:8080/#) in connect(self)
    918         # need this to get the version of weaviate for version checks and proper GRPC configuration
    919         try:
--> 920             meta = executor.result(self.get_meta(False))
    921             self._weaviate_version = _ServerVersion.from_string(meta["version"])
    922             if "grpcMaxMessageSize" in meta:

[/usr/local/lib/python3.11/dist-packages/weaviate/connect/v4.py](https://localhost:8080/#) in get_meta(self, check_is_connected)
    885             return data
    886 
--> 887         return executor.execute(
    888             response_callback=resp,
    889             method=self.get,

[/usr/local/lib/python3.11/dist-packages/weaviate/connect/executor.py](https://localhost:8080/#) in execute(method, response_callback, exception_callback, *args, **kwargs)
     97         return resp_call
     98     except Exception as e:
---> 99         return cast(T, exception_callback(e))
    100 
    101 

[/usr/local/lib/python3.11/dist-packages/weaviate/connect/executor.py](https://localhost:8080/#) in raise_exception(e)
     36 
     37 def raise_exception(e: Exception) -> Any:
---> 38     raise e
     39 
     40 

[/usr/local/lib/python3.11/dist-packages/weaviate/connect/executor.py](https://localhost:8080/#) in execute(method, response_callback, exception_callback, *args, **kwargs)
     93 
     94             return _execute()
---> 95         resp_call = response_callback(call)
     96         assert not isinstance(resp_call, Awaitable)
     97         return resp_call

[/usr/local/lib/python3.11/dist-packages/weaviate/connect/v4.py](https://localhost:8080/#) in resp(res)
    881     def get_meta(self, check_is_connected: bool = True) -> executor.Result[Dict[str, str]]:
    882         def resp(res: Response) -> Dict[str, str]:
--> 883             data = _decode_json_response_dict(res, "Meta endpoint")
    884             assert data is not None
    885             return data

[/usr/local/lib/python3.11/dist-packages/weaviate/util.py](https://localhost:8080/#) in _decode_json_response_dict(response, location)
    691             raise ResponseCannotBeDecodedError(location, response)
    692 
--> 693     raise UnexpectedStatusCodeError(location, response)
    694 
    695 

UnexpectedStatusCodeError: Meta endpoint! Unexpected status code: 404, with response body: None.
```

my assumption here is because we are defining some parameters and setting them to X values while they shouldn't be passed (example port is set to 8080 when it should be None, checkout the default signature for the `weaviate.connect_to_custom` ) also for future this script takes into account a url that looks something like this `http://localhost:8080/` while my url is `abcdefghijk.c0.us-west3.gcp.weaviate.cloud`.

try to double down on this section and use the snipped above as a reference

File: src/chonkie/friends/handshakes/weaviate.py [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+        else:
+            self.client = client
```
Comment (chonk-lain) [inline src/chonkie/friends/handshakes/weaviate.py:130]: loved this 🦛

File: src/chonkie/friends/handshakes/weaviate.py [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+                # Check if the collection exists
+                if not self._collection_exists(self.collection_name):
+                    break
+            print(f"🦛 Chonkie created a new collection in Weaviate: {self.collection_name}")
+        else:
+            self.collection_name = collection_name
+            
+        # Create the collection if it doesn't exist
+        if not self._collection_exists(self.collection_name):
+            self._create_collection()
```
Comment (chonk-lain) [inline src/chonkie/friends/handshakes/weaviate.py]: nice :)

**Review by chonk-lain — COMMENTED** [review]
Thanks a lot for adding this @xucailiang 🙌
could you: 
* format your files using ruff
* attend to the edge case as highlighted below, if you want to test things in colab use the following code snippet as a reference 
```
!pip install -q git+https://github.com/chonkie-inc/chonkie.git@refs/pull/204/head#egg=chonkie[weaviate-client]
!pip install -qU datasets weaviate-client protobuff
```
```python
from chonkie import SemanticChunker, WeaviateHandshake

# METHOD 1

# import weaviate
# from weaviate.classes.init import Auth
# client = weaviate.connect_to_weaviate_cloud(
#     cluster_url=weaviate_url,
#     auth_credentials=Auth.api_key(weaviate_api_key),
# )
# handshake = WeaviateHandshake(client=client)

# METHOD 2 

handshake = WeaviateHandshake(url, api_key)

chunker = SemanticChunker( )
chunks = chunker.chunk(sample_text)

handshake.write(chunks)
```
* update the pyproject.toml file

File: src/chonkie/friends/handshakes/weaviate.py (hunk 1 above)
Comment (xucailiang) [reply src/chonkie/friends/handshakes/weaviate.py]: Right, sorry about that, I forgot.

**Comment by xucailiang** [comment]
> Thanks a lot for adding this @xucailiang 🙌 could you:
> 
> * update the pyproject.toml file
> * checkout the failing test at https://github.com/chonkie-inc/chonkie/actions/runs/16796338114/job/47636554246?pr=204 (other failing tests are unrelated)
> * attend to the edge case as highlighted below, if you want to test things in colab use the following code snippet as a reference
> 
> ```
> !pip install -q git+https://github.com/chonkie-inc/chonkie.git@refs/pull/204/head#egg=chonkie[weaviate-client]
> !pip install -qU datasets weaviate-client protobuff
> ```
> 
> ```python
> from chonkie import SemanticChunker, WeaviateHandshake
> 
> # METHOD 1
> 
> # import weaviate
> # from weaviate.classes.init import Auth
> # client = weaviate.connect_to_weaviate_cloud(
> #     cluster_url=weaviate_url,
> #     auth_credentials=Auth.api_key(weaviate_api_key),
> # )
> # handshake = WeaviateHandshake(client=client)
> 
> # METHOD 2 
> 
> handshake = WeaviateHandshake(url, api_key)
> 
> chunker = SemanticChunker( )
> chunks = chunker.chunk(sample_text)
> 
> handshake.write(chunks)
> ```

Do you mean that `chunker.chunk(sample_text)` will return different result types? 

I noticed that in the latest changes, the `__call__()` parameters were updated to include `Sequence[Sequence[Chunk]]`. [base.py](https://github.com/chonkie-inc/chonkie/blob/main/src/chonkie/chunker/base.py#L33)

So I wrote a simple converter in utils. [utils](https://github.com/xucailiang/chonkie/blob/main/src/chonkie/friends/handshakes/utils.py#L49) 

I'm not sure if this is the best way to do it. If you think it's okay, I can add this converter to the other handshake classes as well.

File: src/chonkie/friends/handshakes/weaviate.py (hunk 2 above)
Comment (xucailiang) [reply src/chonkie/friends/handshakes/weaviate.py]: Great question! The port value shouldn't be None. 

Also,I've modified how WeaviateHandshake is initialized by integrating `weaviate.connect_to_weaviate_cloud()` directly into the initializer, which should reduce the burden on the user.

**Comment by chonk-lain** [comment]
Thanks a lot for the updates 🙌 
I'll try to help out with some commits if you don't mind

**Comment by xucailiang** [comment]
> Thanks a lot for the updates 🙌 I'll try to help out with some commits if you don't mind

Yes plz,feel free

**Review by chonk-lain — APPROVED** [review]
